### ISSUE TITLE
Anonymous Function Unparsable Section

### ISSUE THREAD

**Issue by mariugul** [sqlfluff/sqlfluff#2018]
## Expected Behaviour
Using anonymous DO functions does not work. Unparsable section is outputed. These functions are pgsql functions and should be treated just like CREATE FUNCTION where the language is pgsql.

[Documentation](https://www.postgresql.org/docs/14/sql-do.html)

## Observed Behaviour
```bash
$ sqlfluff lint --dialect postgres do.sql 
== [do.sql] FAIL                                                                                                                                             
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'DO
                       | $$DECLARE r record;\nBEGIN\n FOR r I...'
All Finished 📜 🎉!
```
## Steps to Reproduce
**do.sql**:
```sql
DO $$DECLARE r record;
BEGIN
    FOR r IN SELECT table_schema, table_name FROM information_schema.tables
             WHERE table_type = 'VIEW' AND table_schema = 'public'
    LOOP
        EXECUTE 'GRANT ALL ON ' || quote_ident(r.table_schema) || '.' || quote_ident(r.table_name) || ' TO webuser';
    END LOOP;
END$$;
```

`sqlfluff lint --dialect postgres do.sql `

## Dialect
`postgres`

## Version
`sqlfluff, version 0.8.2`
`Python 3.8.10`
`pip 20.0.2`

## Configuration
`--dialect postgres`

**Comment by PLBMR** [comment]
Looks like bug still exists as of `sqlfluff==0.9.1`:

```bash
mkdir -p ~/.virtualenvs
python3.8 -m venv ~/.virtualenvs/issue-2018-reproduce
source ~/.virtualenvs/issue-2018-reproduce/bin/activate
pip install pip==20.0.2 sqlfluff==0.9.1
...
sqlfluff lint --dialect postgres do.sql
== [do.sql] FAIL                                                                                                                                                                  
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'DO
                       | $$DECLARE r record;\nBEGIN\n FOR r I...'
All Finished 📜 🎉!
```

I will work on addressing this bug in [this branch](https://github.com/PLBMR/sqlfluff/tree/plbmr-issue_2018_anonymous_function_unparsable_section).
